semgrep/semgrep/semgrep_types.py

```diff
@@ -118,9 +119,42 @@ def pattern_names_for_operators(operators: List[Operator]) -> List[str]:
 class Range(NamedTuple):
     start: int
     end: int
+    vars: Mapping[str, Any]
 
     def is_enclosing_or_eq(self, other_range: "Range") -> bool:
-        return self.start <= other_range.start and other_range.end <= self.end
+        return (
+            self.start <= other_range.start
+            and other_range.end <= self.end
+            and self.vars_match(other_range)
+        )
+
+    def vars_match(self, rhs: "Range") -> bool:
+        """
+        Returns true if and only if all metavariables in both this and the other Range refer to the same
+        variables (if variable nodes), in the same scope, or the same expressions (if expression nodes).
+
+        That is, if two patterns define a "$X", and $X refers to a variable in one pattern, then
+        $X must refer to the same variable in both patterns
+        :param rhs: The other Range
+        """
+        to_match = set(self.vars.keys()).intersection(rhs.vars.keys())
+        return all(
+            v in self.vars and v in rhs.vars and self.vars[v] == rhs.vars[v]
```

Comment:
Do we need to do a membership check if we are looping over a set intersection?

Reply:
oh yeah that's stale